docs/release-process.md

```diff
@@ -0,0 +1,178 @@
+  - `latest` in `major.x` branch.
+  - `previous` in `major-1.x` branch.
+  - `life-support` in `major-2.x` branch.
+
+Minor and patch releases of these versions are released from these branches and
+updates are backported from the `master`. Merges to the released version
+branches must happen via Pull Requests to give other maintainers and reviewers
+time to review and verify what will be released.
+
+Process below will be described as procedurs for the situations:
+  - release `next` major version candidate.
+  - release `next` major version. (becomes `latest`)
+  - release `latest` minor and patch versions.
+  - release `previous` minor and patch versions.
+  - release `life-support` minor and patch versions.
+
+Process for the `latest` and `previous` minor/patch releases is the same.
+At the same time only these branches should exist in the repo
+(Example we have just released `v4.0.0`):
+  - `master` - active developmeny
+  - `v4.x` - just become `latest` version.
+  - `v4.x-proposal` - active backport and PR base.
+  - `v3.x` - just become `previous` version.
+  - `v3.x-proposal` - import minor/patch backport and PR base.
+  - `v2.x` - just become `life-support` version.
+  - `v2` does not have proposal, because it only supports critical fixes, which
+    wont go through standard PR release process.
+
+### master branch and package.json version
+  `master` branch only accumulates changes from PRs and is mostly untested(other
+than PR testing). `next` release candidate will branch off from the master.
+Package version in the `master` branch will always be set to the `next` but
+will have `patch` version set to `99` to signify the unsafe version
+(`latestMajor+1.0.99`).
+
+### Release candidate and major release
+  `master` branch already contains everything that needs to be released
+with the `next` version. Instead of directly creating branch to the
+`next` version, we create `branch-proposal` to allow brief review
+of the contents of the release. The purpose of proposal branches is to allow
+other maintainers and reviewers to suggest including something else.
+Proposal should not get dragged out and take max 2-3 days before moving on.
+At this point we don't want to add more complex things to the release candidate,
+only `patch`es until release(feature lock).
+
+Process example (e.g. `latest` was `3.1.0` and we are releasing `4.0.0-rc.1`):
+  - create branch `4.x-proposal` from the `master`
+  - update `package.json` version to `5.0.99` in `master` branch.
+  - update `package.json` version to `4.0.0-rc.1` in `4.x-proposal` branch.
+  - optional: In case we want to omit some changes from the `master`,
+    you can also rebase and discard those commits.
+  - Create Release doc file for the
+    `docs/release-notes/release-notes-4.0.0-draft.md`.
+  - Update CHANGELOG file with the incompatibilities and actions to take.
+  - Create PR: `Release v4.0.0-rc.1` from branch `4.x-proposal` to `4.x`.
```

Comment:
Sure, I may have missed that in the doc. For the simplicity I believe `4.x` will be created before `4.x-proposal` is created from the `master`. As the point of this part (when creating new major release candidate) is to check the list that got included, make sure CHANGELOG/Release notes are updated properly as well as versions and other minor bits are correctly adjusted.

So `4.x` will branch of from the `master` before `4.x-proposal`.

Initially I thought it should branch of from old branches (`3.x` branch), but I doubt this will give us any useful benefits. It will only make rebase harder on the master.